src/libraries/Microsoft.Extensions.HostFactoryResolver/src/HostFactoryResolver.cs

```diff
@@ -112,5 +139,184 @@ private static bool IsFactory<TReturn>(MethodInfo? factory)
             var servicesProperty = hostType.GetProperty("Services", DeclaredOnlyLookup);
             return (IServiceProvider?)servicesProperty?.GetValue(host);
         }
+
+        // This host builder captures calls to the IHostBuilder then replays them on the application's
+        // IHostBuilder when the event fires
+        private class DeferredHostBuilder : IHostBuilder, IObserver<DiagnosticListener>, IObserver<KeyValuePair<string, object?>>
+        {
+            public IDictionary<object, object> Properties { get; } = new Dictionary<object, object>();
+
+            private readonly string[] _args;
+            private readonly MethodInfo _entryPoint;
+
+            private readonly TaskCompletionSource<IHost> _hostTcs = new();
+            private IDisposable? _disposable;
+
+            private Action<IHostBuilder> _configure;
+
+            // The amount of time we wait for the diagnostic source events to fire
+            private static readonly TimeSpan _waitTimeout = TimeSpan.FromSeconds(20);
+
+            public DeferredHostBuilder(string[] args, MethodInfo entryPoint)
+            {
+                _args = args;
+                _entryPoint = entryPoint;
+                _configure = b =>
```

Comment:
```suggestion
                _configure = innerBuilder =>
```

or something similar. To make it easier to understand what this variable is.